### PR TITLE
KubeletPluginsWatcher feature is beta in 1.12 release

### PR DESCRIPTION
**Pull request by RenaudWasTaken**
*What this PR does / why we need it:*
Graduates DevicePlugins feature to beta.

*Which issue(s) this PR fixes:*
Related but does not fix: https://github.com/kubernetes/features/issues/595 as well as https://github.com/kubernetes/kubernetes/issues/65773

*Special notes for your reviewer:*
Includes upgrading the gRPC pluginwatcher API to beta. Based on the [device plugin model](https://github.com/kubernetes/kubernetes/pull/59588).

*Depends on https://github.com/kubernetes/kubernetes/pull/64621 being merged* 

Release note:

```release-note
KubeletPluginsWatcher feature graduates to beta.
```

/sig node
/sig storage

/cc @vladimirvivien @sbezverk @vikaschoudhary16 @saad-ali @vishh @jiayingz

### CODE REVIEW

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

File: pkg/kubelet/apis/pluginregistration/v1beta1/api.proto [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	// For DevicePlugin, this is the resource name that the plugin manages and
+	// should follow the extended resource name convention.
+	// For CSI, this is the CSI driver registrar name.
+	string name = 2;
+        // Optional endpoint location. If found set by Kubelet component,
+        // Kubelet component will use this endpoint for specific requests.
+        // This allows the plugin to register using one endpoint and possibly use
+        // a different socket for control operations. CSI uses this model to delegate
+        // its registration external from the plugin.
+        string endpoint = 3;
```
Comment (vishh) [inline pkg/kubelet/apis/pluginregistration/v1beta1/api.proto]: nit: Fix indentation.

**Review by vishh — COMMENTED** [review]
Can you summarize the level of test coverage that we have for this feature? 
How confident are we with the API?

**Comment by RenaudWasTaken** [comment]
PR updated to fix the indentation.

@vishh 
> Can you summarize the level of test coverage that we have for this feature?
> How confident are we with the API?

- 80.6% go test coverage
- We have unit test coverage of this feature on the device plugin side
- I believe we have e2e node test coverage of this feature
- We have an in flight PR on the device plugin side for e2e tests

This is also the feature that CSI uses for it's plugins. 

IMHO  I'm confident the API won't need to be changed.

**Comment by RenaudWasTaken** [comment]
/retest 

@vishh > Alternatively, you can summarize the test scenarios covered and briefly describe how the existing test coverage could be considered in place of real world testing.

After some offline discussion with Vishh around the level of testing that was put in place,
below is a description of what we have done on this.

As mentioned above, we have fairly good coverage of the code both through unit tests as well as e2e_node tests. All the node device plugin tests have been updated to test the pluginwatcher path.
Here are the scenarios that are covered in go tests and e2e_node:
- [Go][E2E] Regular "Happy" path: Registers 1, 2, 3 device plugins
   -> GO: https://github.com/kubernetes/kubernetes/blob/d13242b72994e32bc13bdc5ae11b0e6ae59c810e/pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go#L59-L83
   -> E2E: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin.go#L66-L93
- [Go][E2E] Kubelet restart, confirms that the plugin is re-registered
   -> E2E: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin.go#L112-L137
   -> GO: https://github.com/kubernetes/kubernetes/blob/d13242b72994e32bc13bdc5ae11b0e6ae59c810e/pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go#L120-L166
- [Go][E2E] Plugin restart, confirms that the plugin is re-registered
- [Go][E2E] Plugin upgrade, confirms that the plugin is re-registered
   -> E2E: https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/device_plugin.go#L167-L181
   -> Go: https://github.com/kubernetes/kubernetes/blob/d13242b72994e32bc13bdc5ae11b0e6ae59c810e/pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go#L85-L118

Additionally a few error paths are tested in the go tests, such as unsupported versions, plugin returning an error during the registration lifecycle, ...

Additionally I've created a new branch for the NVIDIA Device Plugin to use that feature. And it greatly simplifies the code path for plugin developpers: https://github.com/RenaudWasTaken/k8s-device-plugin/tree/v1.12-pluginwatcher
This has been tested on a small cluster (3 nodes) internally.

Though I'm afraid we won't get much production experience until that feature is enabled by default.

**Comment by vishh** [comment]
Based on offline discussions with @RenaudWasTaken and @saad-ali I'm approving this PR. Thanks for the testing summary @RenaudWasTaken. For the future, it would be great if alpha features get some feedback from end users prior to graduating to beta. 

/approve

**Comment by verult** [comment]
@RenaudWasTaken the CSI test failure is likely due to missing driver-registrar params, so CSI driver isn't getting registered by the kubelet registration mechanism. Try this:
https://github.com/verult/kubernetes/commit/666eff4267ec5d0ec7f4242f27415e5a331dfe18

**Comment by RenaudWasTaken** [comment]
@verult thanks! I added pulled your patch on top

**Review by vladimirvivien — COMMENTED** [review]
I lgtm'd it
